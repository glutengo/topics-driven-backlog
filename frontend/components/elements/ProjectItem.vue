<template>
    <li class="project-item card">
        <div 
            v-if="view === 'manage'" 
            class="card-body">

            <template v-if="!editing">
                <BButton 
                    size="sm" 
                    variant="danger" 
                    class="float-right ml-2" 
                    v-confirm="deleteProject">Delete</BButton>

                <BButton 
                    size="sm" 
                    variant="primary" 
                    class="float-right" 
                    @click="startEditing">Edit</BButton>
                
                <h3 class="card-title h5">{{ data.title }}</h3>
            </template>

            <ProjectForm 
                v-else 
                v-model="editingData" 
                @cancel="editing = false" 
                @submit="saveProject"/>
        </div>

        <div v-else-if="view === 'new'" :class="$style.full">
            <div :class="$style.full" v-if="!editing" @click="startEditing">
                <VIcon name="plus" scale="2"/>
            </div>

            <ProjectForm
                v-else
                v-model="editingData"
                @cancel="editing = false"
                @submit="emitProject"/>
        </div>

        <div 
            v-else 
            :class="['card-body', $style.display]">

            <h3 class="card-title h4">
                <router-link :to="baseUrl">{{ data.title }}</router-link>
            </h3>

            <p class="card-text">
                <router-link 
                    class="btn btn-primary mr-2 mb-2"
                    :to="backlogUrl">Backlog</router-link>

                <router-link
                    class="btn btn-primary mr-2 mb-2"
                    :to="sprintPlanningUrl">Sprint planning</router-link>
            </p>
        </div> 
    </li>
</template>

<script>
import '@icons/plus';

import * as _ from 'lodash';
import VIcon from 'vue-awesome/components/Icon';
import BButton from '@bootstrap/button/button';
import ProjectForm from '@/components/forms/ProjectForm';

export default {
    name: 'ProjectItem',
    components: { ProjectForm, BButton, VIcon },
    props: {
        data: {
            type: Object,
            default: () => {},
        },
        view: {
            type: String,
            default: 'manage',
        },
    },
    data() {
        return {
            editing: false,
            editingData: {},
        };
    },
    computed: {
        course() {
            console.log(this.$store.getters);
            return this.$store.getters['courses/byId'](this.data.course_id);
        },

        baseUrl() {
            return '/courses'
                + `/${this.data.course_id}-${this.slugify(this.course.title)}`
                + '/projects'
                + `/${this.data.id}-${this.slugify(this.data.title)}`;
        },

        backlogUrl() {
            return `${this.baseUrl}#backlog`;
        },

        sprintPlanningUrl() {
            return `${this.baseUrl}#sprint-planning`;
        },
    },
    methods: {
        /**
         * Start editing process of this project
         */
        startEditing() {
            if (!this.editing){
                this.editing = true;
                this.editingData = _.pick(this.data, [
                    'title',
                ]);
            }
        },

        /**
         * Save the edited parameters of this project
         */
        async saveProject() {
            await this.$store.dispatch('projects/update', {
                id: this.data.id,
                parentId: this.data.course_id,
                ...this.editingData,
            });

            this.editing = false;

            // TODO handle errors in UI
        },

        /**
         * Delete this project
         */
        async deleteProject() {
            await this.$store.dispatch('projects/remove', {
                id: this.data.id,
                parentId: this.data.course_id,
            });

            // TODO handle errors in UI
        },

        emitProject() {
            this.$emit('submit', this.editingData);
            this.editing = false;
        },

        slugify(text) {
            return text.toString().toLowerCase()
                .replace(/\s+/g, '-')           // Replace spaces with -
                .replace(/[^\w\-]+/g, '')       // Remove all non-word chars
                .replace(/\-\-+/g, '-')         // Replace multiple - with single -
                .replace(/^-+/, '')             // Trim - from start of text
                .replace(/-+$/, '');            // Trim - from end of text
        },
    },
};
</script>

<style lang="scss" module>
    .display {
        display: flex;
        flex-direction: column;
        justify-content: space-between;
    }

    .full {
        height: 100%;
        width: 100%;
        display: flex;
        align-items: center;
        justify-content: center;

        div {
            cursor: pointer;
        }
    }
</style>