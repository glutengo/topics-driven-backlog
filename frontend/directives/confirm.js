import Vue from 'vue';

/**
 * v-confirm directive, asking the user to confirm the action before
 * a click event on an item is processed
 */
Vue.directive('confirm', {
    bind(el, binding) {
        el.handleClick = () => {
            if (confirm('Are you sure? This action is irreversible.')) {
                binding.value();
            }
        };

        el.addEventListener('click', el.handleClick);
    },
    unbind(el) {
        el.removeEventListener('click', el.handleClick);
    },
});
