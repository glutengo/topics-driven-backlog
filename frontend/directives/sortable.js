import Vue from 'vue';
import * as Sortable from 'sortablejs/Sortable';

/**
 * v-sortable directive, binding Sortable plugin to a list
 */
Vue.directive('sortable', {
    inserted: function(el, binding, vnode) {
        if (binding.value) {
            vnode.context.sortableInstance = Sortable.create(el, binding.value);
        }
    },
});
