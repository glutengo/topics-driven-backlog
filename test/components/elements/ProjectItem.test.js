import { mount } from 'vue-test-utils';
import ProjectItem from '@/components/elements/ProjectItem';
import '@/directives/confirm';


describe('The ProjectItem component', () => {
    let cmp, dispatch = jest.fn(), form;

    const mocks = {
        $store: {
            state: {
                user: { id: 1 },
            },  
            dispatch,
            getters: {
                'courses/byId': () => {
                    return {
                        id: 1,
                        allow_enrollment: true,
                    };
                },
                'projects/all': () => {
                    return [];
                },  
            },
        },
        $style: {},
    };


    beforeAll(() => {
        cmp = mount(ProjectItem, {
            propsData: {
                data: {
                    id: 1,
                    title: 'Test Project',
                    user_ids: [],
                },
            },
            mocks,
        });
    });

    beforeEach(() => {
        jest.resetAllMocks();
    });

    it('is a list item that contains a card body', () => {
        expect(cmp.is('li.card')).toBe(true);
        expect(cmp.findAll('.card-body').length).toBe(1);
    });

    it('has a the project title as title', () => {
        const title = cmp.find('.card-title');
        expect(title.text()).toBe('Test Project');
    });

    it('shows join project button and dispatches enroll action when clicked', () => {
        const joinButton = cmp.find('.card-text button.btn-primary');
        expect(joinButton.text()).toBe('Join project');

        joinButton.trigger('click');
        expect(dispatch).toHaveBeenCalled();
    });

    describe('in "new" mode', () => {
        let cmp;

        beforeAll(() => {
            cmp = mount(ProjectItem, {
                propsData: {
                    data: {
                        id: 1,
                        title: 'Test Project',
                        user_ids: [],
                    },
                    view: 'new',
                    courseId: 1,
                },
                mocks,
            });
        });

        it('shows a plus icon and text', () => {
            expect(cmp.find('.fa-icon')).toBeTruthy();
            expect(cmp.find('strong').text()).toBe('Create new project in course');
        });

        it('shows project form when plus is clicked', () => {
            cmp.find('div div').trigger('click');
            form = cmp.find('.project-form');
            expect(form).toBeTruthy();
        });

        it('dispatches project create action', () => {
            form.trigger('submit');

            expect(dispatch).toHaveBeenCalledWith('projects/create', {
                parentId: 1,
                title: 'Test Project',
            });
        });
    });

    describe('when the user is enrolled', () => {
        beforeAll(() => {
            cmp = mount(ProjectItem, {
                propsData: {
                    data: {
                        id: 1,
                        course_id: 1,
                        title: 'Test Project',
                        user_ids: [1],
                    },
                },
                mocks,
            });
        });

        beforeEach(() => {
            jest.resetAllMocks();
        });

        it('shows a link to the backlog', () => {
            const link = cmp.find('.card-text router-link');
            expect(link.text()).toBe('View Backlog');
            expect(link.vnode.data.attrs.to).toBe('/projects/1-test-project');
        });

        it('shows leave project button and dispatches disenroll action when clicked', () => {
            const confirmSpy = jest
                .spyOn(window, 'confirm')
                .mockImplementation(() => true);

            const leaveButton = cmp.find('.card-text button.btn-primary');
            expect(leaveButton.text()).toBe('Leave project');

            leaveButton.trigger('click');
            expect(confirmSpy).toHaveBeenCalled();
            expect(dispatch).toHaveBeenCalled();
        });

        it('shows project form when edit button is clicked', () => {
            const editButton = cmp.find('.btn-outline-primary');
            expect(editButton.find('.fa-icon')).toBeTruthy();
            editButton.trigger('click');

            form = cmp.find('.project-form');
            expect(form).toBeTruthy();
        });
    
        it('dispatches save action when form is submitted', () => {
            form.trigger('submit');

            expect(dispatch).toHaveBeenCalledWith('projects/update', {
                id: 1,
                parentId: 1,
                title: 'Test Project',
            });
        });

        it('dispatches delete action when delete button is clicked', () => {
            const confirmSpy = jest
                .spyOn(window, 'confirm')
                .mockImplementation(() => true);
    
            cmp.update();
    
            const deleteButton = cmp.find('.btn-outline-danger');
            expect(deleteButton.find('.fa-icon')).toBeTruthy();
            deleteButton.trigger('click');
            
            expect(confirmSpy).toHaveBeenCalled();
            expect(dispatch).toHaveBeenCalledWith('projects/remove', {
                id: 1,
                parentId: 1,
            });
        });

        describe('and enrollment is closed', () => {
            beforeAll(() => {
                cmp = mount(ProjectItem, {
                    propsData: {
                        data: {
                            id: 1,
                            course_id: 1,
                            title: 'Test Project',
                            user_ids: [1],
                        },
                    },
                    mocks,
                });

                cmp.vm.$store.getters['courses/byId'] = () => {
                    return {
                        id: 1,
                        allow_enrollment: false,
                    };
                };

                cmp.update();
            });
    
            beforeEach(() => {
                jest.resetAllMocks();
            });

            it('does not show leave project button', () => {
                expect(cmp.findAll('.card-text button.btn-primary').length).toBe(0);
            });

            it('does not edit and delete buttons', () => {
                expect(cmp.findAll('.btn-outline-primary').length).toBe(0);
                expect(cmp.findAll('.btn-outline-danger').length).toBe(0);
            });
        });

        describe('into a different project', () => {
            beforeAll(() => {
                cmp = mount(ProjectItem, {
                    propsData: {
                        data: {
                            id: 1,
                            course_id: 1,
                            title: 'Test Project',
                            user_ids: [],
                        },
                    },
                    mocks,
                });

                cmp.vm.$store.getters['projects/all'] = () => {
                    return [{
                        user_ids: [1],
                    }];
                };

                cmp.update();
            });
    
            beforeEach(() => {
                jest.resetAllMocks();
            });

            it('does not show join project button', () => {
                expect(cmp.findAll('.card-text button.btn-primary').length).toBe(0);
            });
        });
    });
});
