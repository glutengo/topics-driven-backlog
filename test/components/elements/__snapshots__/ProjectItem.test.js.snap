// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectItem.test.js is in edit mode after startEditing was called 1`] = `
<li
  class="project-item card"
>
  <div
    class="card-body"
  >
    <form
      class="project-form"
    >
      <div
        aria-labelledby="__BVID__15___BV_label_"
        class="b-form-group form-group"
        id="__BVID__15_"
        role="group"
      >
        <label
          class="col-form-label pt-0"
          for="project-title"
          id="__BVID__15___BV_label_"
        >
          Title
        </label>
        <div
          class=""
        >
          <input
            aria-required="true"
            class="form-control"
            id="project-title"
            required="required"
            type="text"
          />
          <!---->
          <!---->
          <!---->
        </div>
      </div>
       
      <button
        class="btn btn-primary"
        type="submit"
      >
        Save
      </button>
       
      <button
        class="btn btn-grey"
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</li>
`;

exports[`ProjectItem.test.js shows edit and delete buttons 1`] = `
<li
  class="project-item card"
>
  <div
    class="card-body"
  >
    <button
      class="btn float-right ml-2 btn-danger btn-sm"
      type="button"
    >
      Delete
    </button>
     
    <button
      class="btn float-right btn-primary btn-sm"
      type="button"
    >
      Edit
    </button>
     
    <h3
      class="card-title h5"
    >
      Test Project
    </h3>
  </div>
</li>
`;

exports[`The ProjectItem component shows edit and delete buttons 1`] = `
<li
  class="project-item card"
>
  <div
    class="card-body"
  >
    <button
      class="btn float-right ml-2 btn-danger btn-sm"
      type="button"
    >
      Delete
    </button>
     
    <button
      class="btn float-right btn-primary btn-sm"
      type="button"
    >
      Edit
    </button>
     
    <h3
      class="card-title h5"
    >
      Test Project
    </h3>
  </div>
</li>
`;
